Add DELETE action for api/v1/users endpoint

bolatovumar opened the pull request:
See discussion here for details: https://github.com/btcpayserver/btcpayserver/discussions/2304

Kukks commented on BTCPayServer/Controllers/GreenField/UsersController.cs:
```diff
@@ -165,6 +176,75 @@ public async Task<IActionResult> CreateUser(CreateApplicationUserRequest request
         }
 
+        [HttpDelete("~/api/v1/users/{userId}")]
+        [Authorize(Policy = Policies.CanCreateUser, AuthenticationSchemes = AuthenticationSchemes.GreenfieldAPIKeys)]
```
This permission is definitely wrong.  I think we should create a new policy `CanDeleteUser` and then have the checks in there:
* If user id == current user && (API key permissions has `delete user`)
* if user is admin && (API key permissions has `modify server`)

Take a look at the AuthorizationHandler and accompanying files (follow the policies)

bolatovumar replied:
Yeah, I wasn't sure if this was the right one to use. I will add a new one.

Kukks replied:
Sorry missed this reply.

Yes, that's correct

bolatovumar replied:
@Kukks Ok, there is something weird going on here. If we want any non-admin user to be able to delete themselves then I need to update the permission from `"btcpay.server.candeleteuser"` to `"btcpay.user.candeleteuser"` as `"btcpay.server.*"` are forbidden for assignment to non-admin users (see here: https://github.com/btcpayserver/btcpayserver/blob/bbe1442c284f32a133aafb548756b403a144013e/BTCPayServer/Controllers/ManageController.APIKeys.cs#L439). However, if I update it to `"btcpay.user.candeleteuser"` then something like 
```C# 
var canDeleteUser = (await _authorizationService.AuthorizeAsync(User, null, new PolicyRequirement(Policies.CanDeleteUser))).Succeeded;
``` 
always returns `false` (doesn't matter if user is admin or not and whether the permission is assigned to the API key). I'm not sure why this is. Is there something special about `"btcpay.user.*"` permissions?

It seems that `var canDeleteUser = (await _authorizationService.AuthorizeAsync(User, null, new PolicyRequirement(Policies.CanDeleteUser))).Succeeded;` works for admin users (as in it can return `true` or `false` when appropriate) but not for non-admin users (always returns `false`).

![Screen Shot 2021-04-04 at 5 09 12 PM](https://user-images.githubusercontent.com/1934678/113525308-f4bb8700-9568-11eb-925c-9de5b751e04b.png)
^^^ Always returns `false` for both permissions

![Screen Shot 2021-04-04 at 5 10 19 PM](https://user-images.githubusercontent.com/1934678/113525310-f5ecb400-9568-11eb-82e6-9c870b6345d7.png)
^^^ Returns correct values for both permissions

Kukks commented on BTCPayServer/Controllers/GreenField/UsersController.cs:
```diff
@@ -165,6 +176,75 @@ public async Task<IActionResult> CreateUser(CreateApplicationUserRequest request
+            return Ok();
+        }
+
+        private async Task DeleteUserAndAssociatedData(string userId, ApplicationUser user) 
```
I think there is already a delete action available somewhere in the Admin users list. We should make sure they use the same code (move it a service if need be and use said service here and in that page). That way we can add a delete user action in the account settings later

bolatovumar replied:
Yeah, this code replicates that other code. I can share the code so it's not repeated.

bolatovumar replied:
@Kukks added a user service. Let me know if that's what you wanted or if I got something wrong.

Kukks reviewed (changes requested):
Would also 
* need to be added to the swagger docs
* Need tests

bolatovumar commented:
> Would also
> 
> * need to be added to the swagger docs

Added docs.

> * Need tests

Will add tests once functionality is complete and @Kukks OKs it.

Kukks commented on BTCPayServer/Controllers/ManageController.APIKeys.cs:
```diff
@@ -470,6 +470,7 @@ public class PermissionValueItem
                 {
                     {BTCPayServer.Client.Policies.Unrestricted, ("Unrestricted access", "The app will have unrestricted access to your account.")},
                     {BTCPayServer.Client.Policies.CanCreateUser, ("Create new users", "The app will be able to create new users on this server.")},
+                    {BTCPayServer.Client.Policies.CanDeleteUser, ("Delete users", "The app will be able to delete users on this server.")},
```
```suggestion
                    {BTCPayServer.Client.Policies.CanDeleteUser, ("Delete users", "The app will be able to delete users on this server (if an admin, else only your own user).")},
```

Kukks commented on BTCPayServer/Controllers/GreenField/UsersController.cs:
```diff
@@ -165,6 +179,60 @@ public async Task<IActionResult> CreateUser(CreateApplicationUserRequest request
+
+            var roles = await _userManager.GetRolesAsync(user);
+            // We can safely delete the user if it's not an admin user
+            if (!_userService.IsRoleAdmin(roles))
```
I'm not 100% but i think you can get the Roles the user is in without and db rountrip.   The `User` object has the roles assigned to a user set in the greenfield authorization flow.  I think it's ` User.IsInRole()`

bolatovumar replied:
This won't work because `User.IsInRole()` checks if current user is in a particular role but here we are checking if the user we are trying to delete is an admin.

Kukks replied:
got it 👍

bolatovumar commented on BTCPayServer/wwwroot/swagger/v1/swagger.template.users.json:
```diff
@@ -107,6 +107,48 @@
                 ]
             }
+        },
+        "/api/v1/users/{userId}": {
```
@Kukks `"/api/v1/users/me"` already exists for `GET` requests up above. Or do you mean something else here?

Kukks replied:
Yes, you can do a DELETE with the same route

bolatovumar commented on BTCPayServer/wwwroot/swagger/v1/swagger.template.users.json:
```diff
@@ -107,6 +107,48 @@
+                "security": [
+                    {
+                        "API Key": [
+                            "btcpay.server.cancreateuser"
```
@Kukks updated here: 0a7f52fa0fae7d4f93438e78f306011c7d326eae

bolatovumar commented:
@Kukks does this look reasonable now? I updated the PR by adding the `DELETE` `"~/api/v1/users/me"` endpoint with the constraint of `CanDeleteUser` which any non-admin or admin user can use to delete themselves as long as they are not the only admin user on the service. Same for the `"~/api/v1/users/{userId}"` endpoint except it can only be used by admin users.

If this looks good I will add tests and whatnot.

Kukks commented on BTCPayServer/Controllers/GreenField/UsersController.cs:
```diff
@@ -165,6 +196,57 @@ public async Task<IActionResult> CreateUser(CreateApplicationUserRequest request
+        public async Task<ActionResult<ApplicationUserData>> DeleteUser(string userId)
+        {
+            // Only admins should be allowed to delete users
+            if (!User.IsInRole(Roles.ServerAdmin))
```
use the policy for canmodifyserver to detect if it is an admin instead so that we are consistent with everywhere else

bolatovumar replied:
Addressed in this commit: 49f6eb3b1756fd03c0e6a69ea1a763976a949c10

Kukks commented on BTCPayServer/Controllers/GreenField/UsersController.cs:
```diff
@@ -165,6 +196,57 @@ public async Task<IActionResult> CreateUser(CreateApplicationUserRequest request
+            return Ok();
+        }
+
+        private async Task<Boolean> IsAdmin() 
```
Is this used?

bolatovumar replied:
It's unused. Removed in this commit: ff363f99491050bcf7e4683107aa5ef09731f439

Kukks reviewed (changes requested):
Got one final comment but yes it looks good to move forward with Client + tests

bolatovumar commented:
> Got one final comment but yes it looks good to move forward with Client + tests

I added some basic tests for failure cases here: 08f9cf3a4e8e7ed0fd53a92b8a3a3d27edf36005 but for some reason I can't test the success case. 


This works:

```c#
[Fact(Timeout = TestTimeout)]
[Trait("Integration", "Integration")]
public async Task CanDeleteUsersViaApi()
{
    using (var tester = ServerTester.Create(newDb: true))
    {
        await tester.StartAsync();
        var unauthClient = new BTCPayServerClient(tester.PayTester.ServerUri);
        // Should not be authorized to perform this action
        await AssertHttpError(401,
            async () => await unauthClient.DeleteUser("lol user id"));

        var user = tester.NewAccount();
        user.GrantAccess();
        await user.MakeAdmin();
        var adminClient = await user.CreateClient(Policies.Unrestricted);
        // Should 404 if user doesn't exist
        await AssertHttpError(404,
            async () => await adminClient.DeleteUser("lol user id"));
    }
}
```

But as soon as I attempt to delete a user I get a **405 Method Not Allowed** status code for the final assertion which is strange:

```c#
[Fact(Timeout = TestTimeout)]
[Trait("Integration", "Integration")]
public async Task CanDeleteUsersViaApi()
{
    using (var tester = ServerTester.Create(newDb: true))
    {
        await tester.StartAsync();
        var unauthClient = new BTCPayServerClient(tester.PayTester.ServerUri);
        // Should not be authorized to perform this action
        await AssertHttpError(401,
            async () => await unauthClient.DeleteUser("lol user id"));

        var user = tester.NewAccount();
        user.GrantAccess();
        await user.MakeAdmin();
        var adminClient = await user.CreateClient(Policies.Unrestricted);
        // Should 404 if user doesn't exist
        await AssertHttpError(404,
            async () => await adminClient.DeleteUser("lol user id"));
        
        // Should be able to delete a user successfully
        var userToDelete = tester.NewAccount();
        var response = await adminClient.DeleteUser(userToDelete.UserId);
        Assert.IsType<OkResult>(response); 
    }
}
```

![Screen Shot 2021-04-11 at 11 35 00 AM](https://user-images.githubusercontent.com/1934678/114317200-cfc09a00-9abb-11eb-8472-a219b51d7690.png)

Any idea why this might be?

NicolasDorier commented on BTCPayServer/Controllers/GreenField/UsersController.cs:
```diff
@@ -165,6 +196,36 @@ public async Task<IActionResult> CreateUser(CreateApplicationUserRequest request
+            var user = await _userManager.FindByIdAsync(userId);
+            if (user == null)
+            {
+                return NotFound();
```
Do not use naked NotFound like this, see how I did in for example `InvoiceNotFound`

bolatovumar replied:
@NicolasDorier done in this commit: 192cbc9541f1a77398c8530a43d55ea18b1d50f8

NicolasDorier commented:
@bolatovumar should we put this one in next release?

bolatovumar commented:
> @bolatovumar should we put this one in next release?

@NicolasDorier yeah, if it gets reviewed by then.

NicolasDorier commented:
Code seems OK. I don't really like the `UserService`, but maybe can't do better. Will think a bit, then merge.

bolatovumar commented:
@NicolasDorier alright, let me know if you want me to make any changes. I added `UserService` because @Kukks asked me to add it to reuse code. 

In the meantime I rebased the PR on top of latest `master` branch.

Kukks reviewed (changes requested):
Ok one last thing from my end: There is a missing documentation of the new permission that was added in swagger. If you run the test `EnsureSwaggerPermissionsDocumented ` locally, it will tell you how to fix the docs

Joker11156o commented:
tisdag 1 juni 2021 skrev Andrew Camilleri ***@***.***>:

> *@Kukks* requested changes on this pull request.
>
> Ok one last thing from my end: There is a missing documentation of the new
> permission that was added in swagger. If you run the test
> EnsureSwaggerPermissionsDocumented locally, it will tell you how to fix
> the docs
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/btcpayserver/btcpayserver/pull/2340#pullrequestreview-673016551>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASTEYLX3EKXL2AYSZTPZAELTQTH6FANCNFSM4YYOMMWA>
> .
>

bolatovumar commented:
@Kukks ok updated the description in ca3f97c and tests are passing now.

Kukks reviewed (approved):
Did some small refactoring and improved tests coverage
